core/src/main/java/feast/core/service/SpecService.java

```diff
@@ -324,6 +460,61 @@ public ListStoresResponse listStores(ListStoresRequest.Filter filter) {
     }
   }
 
+  /**
+   * Creates or updates a entity in the repository.
+   *
+   * <p>This function is idempotent. If no changes are detected in the incoming entity's schema,
+   * this method will return the existing entity stored in the repository. If project is not
+   * specified, the entity will be assigned to the 'default' project.
+   *
+   * @param newEntity Entity that will be created or updated.
+   * @param projectName Project namespace of Entity which is to be created/updated
+   */
+  @Transactional
+  public ApplyEntityResponse applyEntity(EntityProto.Entity newEntity, String projectName) {
+    // Autofill default project if not specified
+    if (projectName == null || projectName.isEmpty()) {
+      projectName = Project.DEFAULT_NAME;
+    }
+
+    // Validate incoming entity
+    EntityValidator.validateSpec(newEntity);
+
+    // Find project or create new one if it does not exist
+    Project project = projectRepository.findById(projectName).orElse(new Project(projectName));
+
+    // Ensure that the project retrieved from repository is not archived
+    if (project.isArchived()) {
+      throw new IllegalArgumentException(String.format("Project is archived: %s", projectName));
+    }
+
+    // Retrieve existing Entity
+    EntityV2 entity =
+        entityRepository.findEntityByNameAndProject_Name(
+            newEntity.getSpec().getName(), projectName);
+
+    if (entity == null) {
+      // Create new entity since it doesn't exist
+      newEntity = newEntity.toBuilder().setSpec(newEntity.getSpec()).build();
+      entity = EntityV2.fromProto(newEntity);
+    } else {
+      // If the entity remains unchanged, we do nothing.
+      if (entity.toProto().getSpec().equals(newEntity.getSpec())) {
```

Comment:
Does this work for labels as well? So maps will be equal?

Reply:
Yup, tested by updating labels vs. no updates.